tiatoolbox/utils/image.py

```diff
@@ -20,17 +20,174 @@
+"""
+# Make this immutable / non-writable
+PADDING_TO_BOUNDS.flags.writeable = False
+
+
+def normalise_padding_size(padding):
+    """Normalises padding to be length 4 (left, top, right, bottom).
+
+    Given a scalar value, this is assumed to apply to all sides and
+    therefore repeated for each output (left, right, top, bottom). A
+    length 2 input is assumed to apply the same padding to the
+    left/right and top/bottom.
+
+    Args:
+        padding (int or tuple(int)): Padding to normalise.
+
+    Raises:
+        ValueError: Invalid input size of padding (e.g. length 3).
+        ValueError: Invalid input shape of padding (e.g. 3 dimensional).
+
+    Returns:
+        np.ndarray: Numpy array of length 4 with elements containing
+            padding for left, top, right, bottom.
+    """
+    padding_shape = np.shape(padding)
+    if len(padding_shape) > 1:
+        raise ValueError(
+            "Invalid input padding shape. Must be scalar or 1 dimentional."
+        )
+    padding_size = np.size(padding)
+    if padding_size == 3:
+        raise ValueError("Padding has invalid size 3. Valid sizes are 1, 2, or 4.")
+
+    if padding_size == 1:
+        padding = np.repeat(padding, 4)
+    elif padding_size == 2:
+        padding = np.tile(padding, 2)
+    else:
+        padding = np.array(padding)
+    return padding
+
+
+def make_bounds_size_positive(bounds):
+    """Make bounds have positive size and get horizontal/vertical flip flags.
+
+    Bounds with a negative size in either direction with have the
+    coordinates swapped (e.g. left and right or top and bottom swapped)
+    and a respective horizontal or vertical flip flag set in the
+    output to reflect the swaps which occurred.
+
+    Args:
+        bounds (np.ndarray): Length 4 array-like of bounds.
+
+    Returns:
+        tuple: Three tuple of: positive bounds, horizontal_flip boolean,
+            and vertical_flip boolean.
+    """
+    hflip, vflip = False, False
+    _, (width, height) = bounds2locsize(bounds)
+    if width >= 0 and height >= 0:
+        return bounds, hflip, vflip
+    l, t, r, b = bounds
+    if width < 0:
+        l, r = r, l
+        hflip = True
+    if height < 0:
+        t, b = b, t
+        vflip = True
+    bounds = np.array([l, t, r, b])
+    return (bounds, hflip, vflip)
+
+
+def crop_and_pad_edges(
+    bounds: Tuple[int, int, int, int],
+    max_dimensions: Tuple[int, int],
+    region: np.ndarray,
+    pad_mode: str = "constant",
+    pad_constant_values: Union[int, Tuple] = 0,
+) -> np.ndarray:
+    """Apply padding to areas of a region which are outside max dimensions.
+
+    Applies padding to areas of the image region which have coordinates
+    less than zero or above the width and height in `max_dimensions`.
+    Note that bounds and max_dimensions must given for the same image
+    pyramid level (or more generally resolution e.g. if interpolated
+    between levels or working in other units).
```

Comment:
What is the difference between this as `safe_padded_read` ?

Reply:
Mainly that this is a method to support openslide. Safe_padded_read requires an object to support __get_item__ which openslide does not. An alternative would be to implement a wrapper with this but the cropping behaviour of this function would still be required to implement that at least. If we move to using tifffile eventually this might not be needed.

Reply:
To further clarify the issue with openslide is that it doesn't crop at the edges but pads with black pixels so you have to crop them out before applying the numpy padding to allow for consistent behaviour e.g. other padding modes or padding colours. Notmal array like objects such as numpy arrays and glymur images don't require this as they will return a cropped image pas the boundaries and it works out a bit differently.

Reply:
Can we have a [will be deprecated once tiffreader available] as comment message within the code or sthg?